File: problems/0021-merge-two-sorted-lists/solution.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 0021 – Merge Two Sorted Lists.

        Given the heads of two sorted linked lists `list1` and `list2`,
        merge the two lists into one sorted list by splicing together their nodes.

        Return the head of the merged linked list.

        Approach:
        Use a dummy node as a placeholder for the result list head.
        Traverse both input lists simultaneously, and at each step, attach the smaller
        of the two current nodes to the result list.
        Advance the pointer in the list from which the node was taken.
        After one list is exhausted, append the rest of the other list.

        Time Complexity  : O(m + n), where m and n are the lengths of the input lists.
        Space Complexity : O(1), excluding the output list which is constructed in-place.
        """

        dummy = ListNode()
        current = dummy

        while list1 and list2:
            if list1.val < list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next
            current = current.next

        current.next = list1 if list1 else list2

        return dummy.next

File: problems/0021-merge-two-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        /*
        Problem 0021 – Merge Two Sorted Lists.

        Given the heads of two sorted linked lists `list1` and `list2`,
        merge them into one sorted linked list and return the head.

        The merged list should be made by splicing together the nodes of the two lists.

        Approach:
        Use a dummy node to simplify edge-case handling.
        Traverse both lists in parallel, always attaching the smaller node to the tail
        of the merged result. Move forward in the list from which the node was taken.
        Once one list is exhausted, append the remaining part of the other list.

        Time Complexity  : O(m + n), where m and n are the lengths of the input lists.
        Space Complexity : O(1), not counting the output list which is built in-place.
        */

        ListNode dummy(0);
        ListNode* current = &dummy;

        while (list1 != nullptr && list2 != nullptr) {
            if (list1->val < list2->val) {
                current->next = list1;
                list1 = list1->next;
            } else {
                current->next = list2;
                list2 = list2->next;
            }
            current = current->next;
        }

        current->next = (list1 != nullptr) ? list1 : list2;

        return dummy.next;
    }
};

